src/Controls/src/Core/Picker/Picker.cs

```diff
@@ -302,19 +302,21 @@ void CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
 
 		void AddItems(NotifyCollectionChangedEventArgs e)
 		{
-			int index = e.NewStartingIndex < 0 ? Items.Count : e.NewStartingIndex;
+			int insertIndex = e.NewStartingIndex < 0 ? Items.Count : e.NewStartingIndex;
+			int index = insertIndex;
 			foreach (object newItem in e.NewItems)
 				((LockableObservableListWrapper)Items).InternalInsert(index++, GetDisplayMember(newItem));
-			if (index <= SelectedIndex)
+			if (insertIndex <= SelectedIndex)
 				UpdateSelectedItem(SelectedIndex);
 		}
 
 		void RemoveItems(NotifyCollectionChangedEventArgs e)
 		{
-			int index = e.OldStartingIndex < Items.Count ? e.OldStartingIndex : Items.Count;
+			int removeStart = e.OldStartingIndex < Items.Count ? e.OldStartingIndex : Items.Count - e.OldItems.Count;
+			int index = removeStart + e.OldItems.Count - 1;
```

Comment:
This line of code is a bit ambiguous for a quick read:

```cs
int removeStart = (e.OldStartingIndex < Items.Count ? e.OldStartingIndex : Items.Count) - e.OldItems.Count;
```

OR

```cs
int removeStart = e.OldStartingIndex < Items.Count ? e.OldStartingIndex : (Items.Count - e.OldItems.Count);
```

The next line then also does some arithmetic. What is the correct grouping? Should the second bit of arithmetic apply to both branches of the conditional above it?

Reply:
The second grouping is correct, and the math for index should be applied to both branches. I can add some comments if that would be helpful. The `removeStart` line determines the index for where the removed range starts in the `Items` list. For a typical removal notification, the `e.OldStartingIndex` will indicate that. 

However, the existing behavior included removing the last `e.OldItems.Count` from the Items list if the `e.OldStartingIndex` is greater than `Items.Count`. Not sure if that is proper behavior, or if it should just do nothing, but I keep that behavior (except for fixing an off-by-one error) by calculating `Items.Count - e.OldItems.Count` to determine the first item that should be removed from the end in that case.

And then the existing code removed items in reverse order from back to front of the list, so I kept that behavior by initializing the last index to be removed with `removeStart + e.OldItems.Count - 1`.